rlog/2024-12-23-merkle-tree.mdx

```diff
@@ -0,0 +1,213 @@
+
+### Proof of membership
+An interesting and critical property of Merkle trees is their ability to prove that any piece of data is part of its digest.
+This can be done with logarithmic storage and logarithmic computation time.
+
+Suppose that we want to show that data $\ell$ is part of the Merkle tree's digest.
+Additionally, suppose that $\mathsf{hash}$ is the hash function used to construct the tree.
+We assume that the hash function $\mathsf{hash}$ can be computed in constant-time for any input.
+
+Suppose that a prover provides data $\ell$ to a verifier,
+and tells the verifier that $\ell$ corresponds to the $i$th leaf of the Merkle tree.
+For the verifier to be convinced that $\ell$ is part of the digest, he needs to be able to construct the tree's root hash using $\mathsf{hash}$, $i$, $\ell$ and some additional information from the prover.
+Specifically, the prover must provide the sibling hashes for each value that the verifier can compute.
+This enables the verifier to compute the parents of the siblings that the prover provides and the values that he was able to produce himself.
+
+The leaf index $i$ indicates whether a hash value provided by the prover is a left or right sibling.
+This is done by looking at the binary expansion of $i$.
+
+The verifier can compute the leaf $h_0 = \mathsf{hash}(\ell)$.
+Next, using $h_0$'s sibling, $h'_0$, provided by the prover:
+the verifier can compute $h_1 = \mathsf{hash}(h_0 \|h'_0)$ or $h_1 = \mathsf{hash}(h'_0 \| h_0)$
+depending on whether $h'_0$ is a left or right sibling.
+This pathing continues until the verifier either successfully computes the root hash (in $n+1$ hashes) or fails to do so.
+
+The prover has to provide $n$ sibling nodes for the proof of membership.
+
+There is a key detail that is essential for the proof of membership to be secure.
+The root hash has to be provided to the verifier prior to the selection of data $\ell$.
+Otherwise, the prover could generate a series of hash values (with the corresponding root hash) to forge a proof of membership.
+
+#### Capped proof of membership
+Polygon provides an implementation [[4](https://github.com/0xPolygonZero/plonky2/blob/main/plonky2/src/hash/merkle_tree.rs)] of a shortened proof of membership with a slight modification.
+A specific layer of the Merkle tree is published instead of just the root hash.
+By doing this, a capped proof of membership is just the path from leaf to the published layer.
+
+## Extensions of Merkle trees
+
+Merkle trees can be extended in multiple ways.
+In this section, we explore a select few of these extensions.
+
+### Sparse Merkle trees
+
+A sparse Merkle tree (SMT) is a special Merkle tree that can be used to represent digests with nonconsecutive entries.
+Specifically, each digest entry has a particular leaf index.
+For simplicity, we assume that the indexed values are given by the entry's hash.
+We note that this is a sorted SMT.
```

Comment:
"a variant of a sorted SMT" is more precise.

Reply:
For readability, will leave it as sorted SMT.

Reply:
Good.